File: slack-poster/src/main/java/dev/number6/slackposter/model/EntityAttachment.java
package dev.number6.slackposter.model;

import dev.number6.slackposter.ChannelSummaryImage;

import java.text.MessageFormat;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class EntityAttachment extends Attachment {

    public static final String CHANNEL_ENTITY_DETAILS = "{0} ({1})";

    EntityAttachment(ChannelSummaryImage image) {
        this.fallback = "Entities:";
        this.color = "warning";
        this.title = "Entities:";
        List<Field> fields = new ArrayList<>();
        image.getEntityTotals().forEach((k, v) -> {

            String title = asTitleCase(k) + ":";

            String entityLine = v.entrySet().stream()
                    .sorted(Map.Entry.comparingByValue((o1, o2) -> o2 - o1))
                    .map(e -> MessageFormat.format(CHANNEL_ENTITY_DETAILS, e.getKey(), e.getValue()))
                    .collect(Collectors.joining(", "));

            fields.add(new Field(title, entityLine, false));
        });
        this.fields = fields.toArray(new Field[]{});
    }
}

File: slack-poster/src/main/java/dev/number6/slackposter/model/EntityAttachment.kt
package dev.number6.slackposter.model

import dev.number6.slackposter.ChannelSummaryImage
import java.text.MessageFormat
import java.util.*
import java.util.Map.Entry.comparingByValue
import java.util.stream.Collectors

fun String.formatWithEntry(e: Map.Entry<String, Int>): String {
    return MessageFormat.format(this, e.key, e.value)
}

class EntityAttachment(fallback: String, color: String, title: String, fields: Array<Field>) :
        Attachment(fallback = fallback, color = color, title = title, fields = fields) {

    companion object {
        private const val CHANNEL_ENTITY_DETAILS = "{0} ({1})"

        fun fromImage(image: ChannelSummaryImage): EntityAttachment {
            val fallback = "Entities:"
            val color = "warning"
            val title = "Entities:"
            val fields: MutableList<Field> = ArrayList()
            image.entityTotals.forEach { (k, v) ->
                val entryTitle = k.asTitleCase() + ":"
                val entityLine = v.entries.stream()
                        .sorted(comparingByValue { o1, o2 -> o2 - o1 })
                        .map { e -> CHANNEL_ENTITY_DETAILS.formatWithEntry(e) }
                        .collect(Collectors.joining(", "))
                fields.add(Field(entryTitle, entityLine, false))
            }
            return EntityAttachment(fallback = fallback, color = color, title = title, fields = fields.toTypedArray())
        }
    }
}